toolip/fastapi/app.py

```diff
@@ -0,0 +1,61 @@
+from secrets import compare_digest
+
+from fastapi import APIRouter, Depends, FastAPI, HTTPException, status
+from fastapi.openapi.docs import (
+    get_redoc_html,
+    get_swagger_ui_html,
+    get_swagger_ui_oauth2_redirect_html,
+)
+from fastapi.openapi.utils import get_openapi
+from fastapi.responses import HTMLResponse, JSONResponse
+from fastapi.security import HTTPBasic, HTTPBasicCredentials
+
+from .config import conf
+from .constants import DOCS_PATH, OPENAPI_PATH, REDOC_PATH
+
+security = HTTPBasic()
+
+
+def doc_login(credentials: HTTPBasicCredentials = Depends(security)):
+    correct_username = compare_digest(credentials.username, conf.doc_username)
+    correct_password = compare_digest(credentials.password, conf.doc_password)
+    if not (correct_username and correct_password):
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Incorrect username or password",
+            headers={"WWW-Authenticate": "Basic"},
+        )
+    return
+
+
+def create_app_with_basic_auth(prefix: str = '') -> FastAPI:
+    app = FastAPI(docs_url=None, redoc_url=None, openapi_url=None)
```

Comment:
Let's do this differently because we don't want the library to control the creation of the app entirely. Please test that the `join` works in the Exception string:
```suggestion
def create_app_with_basic_auth(app: FastAPI, prefix: str = '') -> None:
    errors = [attr for attr in ['docs_url', 'redoc_url', 'openapi_url'] if getattr(app, attr) is not None]
    if errors:
        raise ValueError(f'The following attributes must be set to None when creating the FastAPI app: {", ".join(errors)})
```